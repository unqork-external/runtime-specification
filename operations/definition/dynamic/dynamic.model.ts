import { DiscriminatorValue, Required } from '@tsed/schema'

import { DynamicOperationOptions } from './dynamic.options'
import { Alpha } from '../../../../decorators'
import { AutogenDisabled } from '../../decorators/autogenDisabled.decorator'
import { OperationTypes } from '../../enums/operation-types.enum'
import { Operation } from '../../interface/operations.interface'

@DiscriminatorValue(OperationTypes.DYNAMIC)
@Alpha()
@AutogenDisabled()
export class DynamicOperation extends Operation<OperationTypes.DYNAMIC> {
  public type = OperationTypes.DYNAMIC as const

  @Required()
  public options: DynamicOperationOptions = new DynamicOperationOptions()
}
