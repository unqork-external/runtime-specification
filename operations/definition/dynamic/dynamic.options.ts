import { CollectionOf, Description, Required } from '@tsed/schema'

import { InputRef } from '../../../inputs/inputRef'
import { OperationOptions } from '../../interface'

export class DynamicOperationOptions extends OperationOptions {
  @Description(
    "Key or path to send operation to. If specified, overrides any `targetKey` in dynamic operation's options",
  )
  targetKey?: string

  @Required()
  @Description('Formula or interpolation string that resolves to a full operation')
  operation: string

  @Description(`
    Inputs to be used within formulas in other operation options.
  `)
  @CollectionOf(InputRef)
  inputs?: InputRef[]
}
