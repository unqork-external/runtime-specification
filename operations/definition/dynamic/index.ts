export * from './dynamic.model'
export * from './dynamic.options'
