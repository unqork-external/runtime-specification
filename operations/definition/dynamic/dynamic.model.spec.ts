import { DynamicOperation } from './dynamic.model'
import { generateSchemaAndValidate } from '../../../../utilities'

describe('Operation Model: DynamicOperation Tests', () => {
  let schema
  beforeAll(() => {
    const response = generateSchemaAndValidate(DynamicOperation)
    schema = response.schema
  })

  it('should match this JSON Schema', () => {
    expect(schema).toMatchSnapshot()
  })
})
