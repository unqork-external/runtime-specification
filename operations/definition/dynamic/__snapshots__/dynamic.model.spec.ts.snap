// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Operation Model: DynamicOperation Tests should match this JSON Schema 1`] = `
{
  "autogenDisabled": true,
  "definitions": {
    "DynamicOperationOptions": {
      "properties": {
        "inputs": {
          "description": "
    Inputs to be used within formulas in other operation options.
  ",
          "items": {
            "$ref": "#/definitions/InputRef",
          },
          "type": "array",
        },
        "operation": {
          "description": "Formula or interpolation string that resolves to a full operation",
          "minLength": 1,
          "type": "string",
        },
        "shouldExecute": {
          "default": true,
          "description": "
    Conditionally execute this operation.
    Can set to false to temporarily disabled when debugging.
  ",
          "type": [
            "string",
            "boolean",
          ],
        },
        "targetKey": {
          "description": "Key or path to send operation to. If specified, overrides any \`targetKey\` in dynamic operation's options",
          "type": "string",
        },
      },
      "required": [
        "operation",
      ],
      "type": "object",
    },
    "InputRef": {
      "description": "Objects may define other targets that they refer to. This becomes a link for operation execution within the Runtime.",
      "properties": {
        "alias": {
          "type": "string",
        },
        "propertyPath": {
          "type": "string",
        },
        "required": {
          "type": "boolean",
        },
        "silent": {
          "type": "boolean",
        },
        "targetKey": {
          "minLength": 1,
          "type": "string",
        },
        "type": {
          "type": "string",
        },
      },
      "required": [
        "targetKey",
      ],
      "type": "object",
    },
  },
  "properties": {
    "creatorSummary": {
      "description": "A detailed summary of the operation",
      "type": "string",
    },
    "name": {
      "description": "Name of the operation",
      "type": "string",
    },
    "options": {
      "$ref": "#/definitions/DynamicOperationOptions",
    },
    "type": {
      "const": "DYNAMIC",
      "examples": [
        "DYNAMIC",
      ],
      "type": "string",
    },
  },
  "required": [
    "options",
  ],
  "type": "object",
}
`;
